### PR TITLE
feat(lint): add args-bounds check for unbounded args[++i] (fixes #1900)

### LINKED ISSUE #1900: lint: flag unbounded args[++i] without bounds check in hand-rolled arg parsers
## Problem

The interrupt handler added in #1898 uses a hand-rolled flag parser:

```ts
if (args[i] === '--reason') {
  rawReason = args[++i];  // UB when --reason is the last arg
}
```

When `--reason` is the last argument, `args[++i]` = `args[args.length]` = `undefined`. The downstream guard (`if (rawReason !== undefined)`) silently skips it — no error, no usage hint. The user thinks a reason was set; it wasn't.

## Proposed fix

Add a lint rule in `scripts/` (simple ripgrep check in CI or pre-commit) that flags any `args[\+\+i]` not immediately preceded by a bounds check (`i + 1 >= args.length`).

Alternatively, enforce use of the existing `extractFullFlag` / `extractJqFlag` helpers from `parse.ts` and disallow raw `args[++i]` outside of that module.

## Short-term workaround (for #1898)

```ts
if (args[i] === '--reason') {
  if (i + 1 >= args.length) {
    d.printError('--reason requires a value');
    d.exit(2);
  }
  rawReason = args[++i];
}
```

Discovered during adversarial review of #1898.

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary

- Adds `scripts/check-args-bounds.ts`: a new lint script (modeled after `check-shell-injection.ts`) that detects `args[++i]` accesses without a bounds guard — the pattern behind the silent `--reason` bug in #1898
- Adds `scripts/check-args-bounds.spec.ts`: 14 unit tests covering all safe and unsafe patterns detected by the `isSafe()` function
- Registers `lint:args-bounds` in `package.json` and adds it to both tiers of the pre-commit hook
- Fixes 8 pre-existing violations across `export.ts`, `remove.ts`, `import.ts` (×2), and `claude.ts` (parsePatchUpdateArgs, parseApproveArgs, parseDenyArgs ×2)

The check classifies a line as safe if any of: `??` on the same line, `args[i + 1]` truthy pre-check, explicit `i + 1 / args.length` bounds check in the preceding 6 lines, or a falsy/null post-check on the assigned variable within the next 2 lines.

## Test plan

- [x] `bun run lint:args-bounds` exits 0 on the patched codebase
- [x] `bun test` (command package) — 2036 pass, 0 fail
- [x] `bun typecheck` — no errors
- [x] `bun run lint:check` — no errors
- [x] Spec: 14/14 unit tests pass (`cd scripts && bun test check-args-bounds.spec.ts`)
- [x] Pre-commit hook ran all checks including the new `lint:args-bounds` step

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by theshadow27** [comment]
## Adversarial Review (Delta)

### Changes Since Last Review

| # | Previous Issue | Status | Notes |
|---|----------------|--------|-------|
| 1 | 🔴 Rule 3 lookback `j < lineIdx` misses same-line ternary | ✅ Fixed in 5d076b6c | Changed to `j <= lineIdx`; added spec case `"same-line ternary bounds check"` |
| 2 | 🔴 Rule 1 `line.includes("??")` matches comments | ✅ Fixed in 5d076b6c | Replaced with `/args\[\+\+i\]\s*\?\?/`; added spec case `"?? in a comment is not a null-coalescing guard"` |
| 3 | 🟡 Rule 2 lookback is 1 line — multi-line `if` false positives | ⏳ Not addressed | Filed as follow-up #1967 |
| 4 | 🟡 No `// lint-allow-args-bounds` escape hatch | ⏳ Not addressed | Filed as follow-up #1968 |
| 5 | 🟡 Rule 3 misses `i < args.length - 1` | ⏳ Not addressed | Filed as follow-up #1969 |
| 6 | 🔵 `new RegExp()` compiled inside loop | ⏳ Not addressed | Minor; not a blocker |

### New Issues (Copilot finding, fixed in 9c307c09)

- ✅ **Rule 3 token-pair test matched comments** — `I_PLUS_1.test(line) && line.includes("args.length")` treated `// TODO: add i + 1 < args.length check` as a real guard. Fixed by replacing both tokens with `BOUNDS_EXPR` (`/i\s*\+\s*1\s*[<>]=?\s*args\.length\b|.../`) and stripping `//` comments before matching. Added spec case `"comment containing i + 1 and args.length is not a bounds guard"`. 19/19 spec cases pass; `lint:args-bounds` still exits 0 on the full codebase.

### Updated Verdict
✅ **Approve** — all 🔴 issues resolved (including the Copilot finding). Three 🟡 follow-ups filed as #1967, #1968, #1969 and out of scope for this PR.

File: scripts/check-args-bounds.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    const lookahead = Math.min(lines.length - 1, lineIdx + 2);
+    for (let j = lineIdx + 1; j <= lookahead; j++) {
+      const next = lines[j];
+      if (
+        next.includes(`!${varName}`) ||
+        next.includes(`${varName} === undefined`) ||
+        next.includes(`${varName} === null`) ||
+        next.includes(`${varName} == null`)
+      ) {
+        return true;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-args-bounds.ts:87]: Fixed in repair commits — post-check now uses word-boundary RegExp (`!\b${varName}\b`) instead of `String.includes()`. Spec case 'post-check matches longer identifier with same suffix (word-boundary)' verifies the fix.

File: scripts/check-args-bounds.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  process.stderr.write("  Bad:   val = args[++i];\n");
+  process.stderr.write("  Good:  if (i + 1 < args.length) val = args[++i];\n");
+  process.stderr.write("  Good:  val = args[++i] ?? defaultValue;\n\n");
+
+  process.exit(1);
+}
+
+main();
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-args-bounds.ts]: Fixed — last line of the file is `if (import.meta.main) main();`, so `main()` does not execute when `isSafe` is imported in the spec.

File: scripts/check-args-bounds.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+export function isSafe(lines: string[], lineIdx: number): boolean {
+  const line = lines[lineIdx];
+
+  // Rule 1: null coalescing on same line
+  if (line.includes("??")) return true;
+
+  // Rule 2: truthy pre-check — current or preceding line contains args[i + 1]
+  if (line.includes("args[i + 1]")) return true;
+  if (lineIdx > 0 && lines[lineIdx - 1].includes("args[i + 1]")) return true;
+
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-args-bounds.ts]: Fixed — Rule 2 now uses `TRUTHY_PRE_CHECK` regex requiring `args[i + 1]` in a boolean context (`&&`, `||`, `if (`, `while (`). Spec case 'bare args[i + 1] read is not a guard' verifies.

File: scripts/check-args-bounds.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  const TRUTHY_PRE_CHECK = /(?:&&|\|\||if\s*\(|while\s*\()[^)]*args\[i\s*\+\s*1\]/;
+  if (TRUTHY_PRE_CHECK.test(line)) return true;
+  if (lineIdx > 0 && TRUTHY_PRE_CHECK.test(lines[lineIdx - 1])) return true;
+
+  // Rule 3: explicit bounds check in the preceding 6 lines *or on the current line*
+  // (includes same-line ternary: `val = i + 1 < args.length ? args[++i] : null`)
+  const lookback = Math.max(0, lineIdx - 6);
+  for (let j = lookback; j <= lineIdx; j++) {
+    if (I_PLUS_1.test(lines[j]) && lines[j].includes("args.length")) return true;
+  }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply scripts/check-args-bounds.ts:73]: Fixed in 9c307c09 — Rule 3 replaced `I_PLUS_1.test() && line.includes('args.length')` with `BOUNDS_EXPR` regex; comment stripping via `.replace(/\/\/.*$/, '')` applied before match. Spec case 'comment containing i + 1 and args.length is not a bounds guard' verifies.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1963 — feat(lint): add args-bounds check for unbounded args[++i]
**Linked issue:** #1900

### Implementation Evidence
- `scripts/check-args-bounds.ts` — new lint script with `isSafe()` heuristic covering 4 rules: null-coalescing (`??`), truthy pre-check (`args[i+1]` in boolean context), explicit bounds comparison (`BOUNDS_EXPR` regex with comment stripping), and post-check on assigned variable (word-boundary RegExp)
- `package.json:27` — `lint:args-bounds` script registered
- `.git-hooks/pre-commit:67,98` — wired into both source-tier and config-only-tier pre-commit paths
- `packages/command/src/commands/claude.ts:1373-1377` — `--source` bounds guard added
- `packages/command/src/commands/claude.ts:1451-1456` — `--request-id` post-check guard added
- `packages/command/src/commands/claude.ts:1488-1493,1496-1500` — `--request-id` and `--message` guards in parseDenyArgs
- `packages/command/src/commands/export.ts:36` — `--scope` bounds guard added
- `packages/command/src/commands/import.ts:242,402` — `--scope` bounds guards added (two loops)
- `packages/command/src/commands/remove.ts:23` — `--scope` bounds guard added

### Test Coverage
- `scripts/check-args-bounds.spec.ts` — 19 focused unit tests for `isSafe()`: 7 violation cases (unsafe), 12 safe cases
- Covers all 4 rules including edge cases: word-boundary post-check, same-line ternary bounds, comment-stripping for Rule 3, bare `args[i+1]` not treated as guard, `??` in comment not treated as null-coalescing guard

### Copilot Threads (Step 5b)
All 4 threads classified and dismissed — all raised against pre-repair code:
- #3171766536 — `String.includes()` false positives → Fixed: `\b` word-boundary RegExp in current code
- #3171766551 — `main()` unconditional → Fixed: `if (import.meta.main) main()` guard present
- #3171766560 — Rule 2 bare substring → Fixed: `TRUTHY_PRE_CHECK` regex requires boolean context
- #3171847079 — Rule 3 comment false positive → Fixed in 9c307c09: `BOUNDS_EXPR` + comment stripping

### Test Results
CI green: `bun test` covered by `check` job (4m1s, pass). 19/19 spec cases confirmed passing per adversarial review delta.

### CI Status
All 3 required checks passing:
- `build`: pass (8s)
- `check`: pass (4m1s)
- `coverage`: pass (6m14s)

### Remaining Work (non-blocking)
- #1967 — Rule 2 lookback is 1 line; multi-line `if` may produce false positives
- #1968 — No `// lint-allow-args-bounds` escape hatch
- #1969 — Rule 3 misses `i < args.length - 1` form

### Outcome
**qa:pass** — All described functionality implemented and verified. Lint script, pre-commit integration, and command-parser fixes all present. 19-case spec suite covers the heuristic thoroughly. CI (build/check/coverage) green.
